Bug: Bucket and entry removing don't work with Azure Blob Fuse

### What happened?

If I try to remove a bucket or entry from a ReductStore instance which uses Azure Blob Fuse, I have the following error:

```
api/middleware.rs:56 DELETE /api/v1/b/datasets_1 [500 Internal Server Error] Directory not empty (os error 39)
```

### Version

v1.3.0

### OS

Linux

### Snippet to repeat the bug

_No response_

### Logs

```Shell
Jan 18 12:44:35 play systemd[1]: Started reductstore.service - ReductStore Service.
Jan 18 12:44:35 play run_reductstore.sh[32339]: + sleep 5
Jan 18 12:44:40 play run_reductstore.sh[32339]: + '[' -z /mnt/data ']'
Jan 18 12:44:40 play run_reductstore.sh[32341]: + blobfuse2 mount list
Jan 18 12:44:40 play run_reductstore.sh[32342]: + grep -q /mnt/data
Jan 18 12:44:40 play blobfuse2[32341]: [] LOG_DEBUG [custom.go (54)]: initializePlugins: No plugins to load.
Jan 18 12:44:40 play run_reductstore.sh[32339]: + /usr/local/bin/reductstore
Jan 18 12:44:40 play run_reductstore.sh[32347]: 2025-01-18 12:44:40.132 (40064) [INFO] -- main.rs:23 ReductStore 1.13.0 [9c85c4f at 2024-12-04T16:36:46Z]
Jan 18 12:44:40 play run_reductstore.sh[32347]: 2025-01-18 12:44:40.133 (40064) [INFO] -- main.rs:38 Configuration:
Jan 18 12:44:40 play run_reductstore.sh[32347]:          RS_LOG_LEVEL = INFO (default)
Jan 18 12:44:40 play run_reductstore.sh[32347]:         RS_HOST = 0.0.0.0 (default)
Jan 18 12:44:40 play run_reductstore.sh[32347]:         RS_PORT = 443
Jan 18 12:44:40 play run_reductstore.sh[32347]:         RS_API_BASE_PATH = / (default)
Jan 18 12:44:40 play run_reductstore.sh[32347]:         RS_DATA_PATH = /mnt/data
Jan 18 12:44:40 play run_reductstore.sh[32347]:         RS_API_TOKEN = ***********
Jan 18 12:44:40 play run_reductstore.sh[32347]:         RS_CERT_PATH = *****************************************
Jan 18 12:44:40 play run_reductstore.sh[32347]:         RS_CERT_KEY_PATH = *****************************************
Jan 18 12:44:40 play run_reductstore.sh[32347]:         RS_LICENSE_PATH = /usr/local/lib/reductstore/license.lic
Jan 18 12:44:40 play run_reductstore.sh[32347]:         RS_CORS_ALLOW_ORIGIN = *
Jan 18 12:44:40 play run_reductstore.sh[32347]:         RS_BUCKET_1_NAME = demo
Jan 18 12:44:40 play run_reductstore.sh[32347]:         RS_BUCKET_2_NAME = datasets
Jan 18 12:44:40 play run_reductstore.sh[32347]:         RS_BUCKET_1_QUOTA_TYPE = FIFO
Jan 18 12:44:40 play run_reductstore.sh[32347]:         RS_BUCKET_1_QUOTA_SIZE = 10.0 TB
Jan 18 12:44:40 play run_reductstore.sh[32347]:         RS_TOKEN_1_NAME = dataset-read
Jan 18 12:44:40 play run_reductstore.sh[32347]:         RS_TOKEN_2_NAME = demo
Jan 18 12:44:40 play run_reductstore.sh[32347]:         RS_TOKEN_1_VALUE = *****************************************************************************
Jan 18 12:44:40 play run_reductstore.sh[32347]:         RS_TOKEN_2_VALUE = ***********
Jan 18 12:44:40 play run_reductstore.sh[32347]:         RS_TOKEN_1_READ = datasets
Jan 18 12:44:40 play run_reductstore.sh[32347]:         RS_TOKEN_2_READ = demo,datasets
Jan 18 12:44:40 play run_reductstore.sh[32347]:         RS_TOKEN_2_WRITE = demo
Jan 18 12:44:41 play run_reductstore.sh[32347]: 2025-01-18 12:44:41.071 (90656) [WARN] -- storage/entry/entry_loader.rs:190 Recovering uncommitted changes from WALs for entry: "/mnt/data/test-6/mnist_training"
Jan 18 12:44:41 play run_reductstore.sh[32347]: 2025-01-18 12:44:41.402 (40064) [INFO] -- storage/storage.rs:67 Load 14 buckets
Jan 18 12:44:41 play run_reductstore.sh[32347]: 2025-01-18 12:44:41.410 (40064) [INFO] -- cfg.rs:146 Provisioned bucket 'demo' with: BucketSettings { quota_type: Some(FIFO), quota_size: Some(10000000000000), max_block_size: Some(64000000), max_block_records: Some(256) }
Jan 18 12:44:41 play run_reductstore.sh[32347]: 2025-01-18 12:44:41.419 (40064) [INFO] -- cfg.rs:146 Provisioned bucket 'datasets' with: BucketSettings { quota_type: Some(NONE), quota_size: Some(0), max_block_size: Some(64000000), max_block_records: Some(256) }
Jan 18 12:44:41 play run_reductstore.sh[32347]: 2025-01-18 12:44:41.424 (40064) [INFO] -- cfg.rs:112 Provisioned token 'dataset-read' with Some(Permissions { full_access: false, read: ["datasets"], write: [""] })
Jan 18 12:44:41 play run_reductstore.sh[32347]: 2025-01-18 12:44:41.424 (40064) [INFO] -- cfg.rs:112 Provisioned token 'demo' with Some(Permissions { full_access: false, read: ["demo", "datasets"], write: ["demo"] })
Jan 18 12:44:41 play run_reductstore.sh[32347]: 2025-01-18 12:44:41.424 (40064) [INFO] -- cfg.rs:372 Load Web Console
Jan 18 12:44:41 play run_reductstore.sh[32347]: 2025-01-18 12:44:41.458 (40064) [INFO] -- main.rs:46 License Information:
Jan 18 12:44:41 play run_reductstore.sh[32347]:         Licensee: Azure MarketPlace
Jan 18 12:44:41 play run_reductstore.sh[32347]:         Invoice: ---
Jan 18 12:44:41 play run_reductstore.sh[32347]:         Expiry Date: 2045-01-01 00:00:00 UTC
Jan 18 12:44:41 play run_reductstore.sh[32347]:         Plan: STANDARD
Jan 18 12:44:41 play run_reductstore.sh[32347]:         Number of Devices: 1
Jan 18 12:44:41 play run_reductstore.sh[32347]:         Disk Quota: 1 TB,
Jan 18 12:44:41 play run_reductstore.sh[32347]:         Fingerprint: 32063a5fd10e4741fa262f8bbc7bb21ce9a01eb57ae34005ec6f50838e718800
Jan 18 12:44:41 play run_reductstore.sh[32347]: 2025-01-18 12:44:41.458 (40064) [INFO] -- main.rs:60 Run HTTP reductstore on https://0.0.0.0:443/
Jan 18 12:45:20 play run_reductstore.sh[32347]: 2025-01-18 12:45:20.707 (40064) [ERROR] -- api/middleware.rs:56 DELETE /api/v1/b/test-6 [500 Internal Server Error] Input/output error (os error 5)
Jan 18 12:45:29 play run_reductstore.sh[32347]: 2025-01-18 12:45:29.444 (40064) [ERROR] -- api/middleware.rs:56 DELETE /api/v1/b/datasets_1 [500 Internal Server Error] Directory not empty (os error 39)
Jan 18 12:45:42 play run_reductstore.sh[32347]: 2025-01-18 12:45:42.715 (40064) [ERROR] -- api/middleware.rs:56 DELETE /api/v1/b/datasets_2 [500 Internal Server Error] Directory not empty (os error 39)
Jan 18 12:51:32 play run_reductstore.sh[32347]: 2025-01-18 12:51:32.655 (40064) [ERROR] -- api/middleware.rs:56 DELETE /api/v1/b/datasets_3 [500 Internal Server Error] Directory not empty (os error 39)
Jan 18 12:57:15 play run_reductstore.sh[32347]: 2025-01-18 12:57:15.787 (40064) [ERROR] -- api/middleware.rs:56 DELETE /api/v1/b/demo/entry_1 [500 Internal Server Error] Directory not empty (os error 39)
Jan 18 13:00:04 play run_reductstore.sh[32347]: 2025-01-18 13:00:04.844 (90656) [ERROR] -- storage/entry/entry_loader.rs:37 [InternalServerError] Block index "/mnt/data/demo/entry_12/blocks.idx" not found
Jan 18 13:00:04 play run_reductstore.sh[32347]: 2025-01-18 13:00:04.844 (90656) [WARN] -- storage/entry/entry_loader.rs:62 Failed to restore from block index. Trying to rebuild the entry from blocks
```

### Code of Conduct

- [x] I agree to follow this project's Code of Conduct

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -10,6 +10,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Fixed
 
 - RS-555: Close WAL file before removal it from disk, [PR-706](https://github.com/reductstore/reductstore/pull/706)
+- RS-583: Remove contents of folder before remove it, [PR-711](https://github.com/reductstore/reductstore/pull/711)
 
 ## [1.13.2] - 2025-01-15
 

--- a/reductstore/src/core/file_cache.rs
+++ b/reductstore/src/core/file_cache.rs
@@ -4,9 +4,9 @@
 use crate::core::cache::Cache;
 use reduct_base::error::ReductError;
 use reduct_base::internal_server_error;
-use std::fs::{remove_dir_all, remove_file, rename, File};
+use std::fs::{read_dir, remove_dir, remove_dir_all, remove_file, rename, File};
 use std::io::{Seek, SeekFrom};
-use std::path::PathBuf;
+use std::path::{Path, PathBuf};
 use std::sync::{Arc, LazyLock, RwLock, Weak};
 use std::time::Duration;
 
@@ -188,7 +188,10 @@ impl FileCache {
 
     pub fn remove_dir(&self, path: &PathBuf) -> Result<(), ReductError> {
         if path.try_exists()? {
-            remove_dir_all(path)?;
+            // Remove all files in the directory recursively
+            // due to some file systems not supporting removing non-empty directories
+            Self::remove_dir_contents(path)?;
+            remove_dir(path)?;
         }
         self.discard_recursive(path)
     }
@@ -221,9 +224,24 @@ impl FileCache {
     ///
     /// A `Result` which is `Ok` if the file was successfully renamed, or an `Err` containing
     pub fn rename(&self, old_path: &PathBuf, new_path: &PathBuf) -> Result<(), ReductError> {
-        rename(old_path, new_path)?;
         let mut cache = self.cache.write()?;
         cache.remove(old_path);
+        rename(old_path, new_path)?;
+        Ok(())
+    }
+
+    fn remove_dir_contents<P: AsRef<Path>>(path: P) -> Result<(), ReductError> {
+        for entry in read_dir(path)? {
+            let entry = entry?;
+            let path = entry.path();
+
+            if entry.file_type()?.is_dir() {
+                Self::remove_dir_contents(&path)?;
+                remove_dir(path)?;
+            } else {
+                remove_file(path)?;
+            }
+        }
         Ok(())
     }
 }

--- a/reductstore/src/core/file_cache.rs
+++ b/reductstore/src/core/file_cache.rs
@@ -4,7 +4,7 @@
 use crate::core::cache::Cache;
 use reduct_base::error::ReductError;
 use reduct_base::internal_server_error;
-use std::fs::{read_dir, remove_dir, remove_dir_all, remove_file, rename, File};
+use std::fs::{read_dir, remove_dir, remove_file, rename, File};
 use std::io::{Seek, SeekFrom};
 use std::path::{Path, PathBuf};
 use std::sync::{Arc, LazyLock, RwLock, Weak};

